File: load-balancer/java/src/main/java/com/danielwetan/softwareengineering/loadbalancer/LoadBalancer.java
package com.danielwetan.softwareengineering.loadbalancer;

import java.util.ArrayList;
import java.util.List;

public class LoadBalancer {
  private static LoadBalancer instance;
  private List<Server> servers;
  private LoadBalancingStrategy strategy;

  private LoadBalancer() {
    this.servers = new ArrayList<>();
  }

  public static LoadBalancer getInstance() {
    if (instance == null) {
      instance = new LoadBalancer();
    }

    return instance;
  }

  public void addServer(Server server) {
    servers.add(server);
  } 

  public void removeServer(Server server) {
    servers.remove(server);
  }

  public Server getServer(Request request) {
    return strategy.getServer(servers, request);
  }

  public void setLoadBalancingStrategy(LoadBalancingStrategy strategy) {
    this.strategy = strategy;
  }
}

File: load-balancer/go/load_balancer.go
package main

import "sync"

type LoadBalancer struct {
	servers  []Server
	strategy LoadBalancingStrategy
}

var instance *LoadBalancer
var once sync.Once

// GetInstance ensures a singleton instance of LoadBalancer.
func GetInstance() *LoadBalancer {
	once.Do(func() {
		instance = &LoadBalancer{
			servers: make([]Server, 0),
		}
	})
	return instance
}

func (lb *LoadBalancer) AddServer(server Server) {
	lb.servers = append(lb.servers, server)
}

func (lb *LoadBalancer) RemoveServer(server Server) {
	for i, s := range lb.servers {
		if s == server {
			lb.servers = append(lb.servers[:i], lb.servers[i+1:]...)
			break
		}
	}
}

func (lb *LoadBalancer) GetServer(request Request) (Server, error) {
	return lb.strategy.GetServer(lb.servers, request)
}

func (lb *LoadBalancer) SetLoadBalancingStrategy(strategy LoadBalancingStrategy) {
	lb.strategy = strategy
}
